batchflow/models/tf/unet.py

```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
+        config = super().build_config(names)
+
+        num_stages = config.get('body/encoder/num_stages')
+
+        if config.get('body/decoder/dense/upsample') is None:
+            config['body/decoder/dense/upsample'] = dict(layout='tna')
+
+        if config.get('body/decoder/dense/filters') is None:
+            config['body/decoder/dense/filters'] = [32 * 2**i for i in range(num_stages)]
+
+        if config.get('body/decoder/dense/layout') is None:
+            config['body/decoder/dense/layout'] = 'cna'
+
+        if config.get('body/decoder/dense/kernel') is None:
+            config['body/decoder/dense/kernel'] = 3
+
+        if config.get('body/decoder/dense/combine_op') is None:
+            config['body/decoder/dense/combine_op'] = 'concat'
+
+        return config
+
+    @classmethod
+    def decoder(cls, inputs, name='decoder', **kwargs):
+        decoder_filters = kwargs['upsample']['filters']
+
+        num_stages = kwargs.pop('num_stages') or len(inputs)-2
+
+        combine_op = kwargs['dense']['combine_op']
+        combine_args = {'op': combine_op if isinstance(combine_op, str) else '',
+                        'data_format': kwargs.get('data_format'),
+                        **(combine_op if isinstance(combine_op, dict) else {})}
+
+        semantic_outputs = []
+        n_outputs = kwargs['n_outputs']
+        if isinstance(n_outputs, int):
+            n_outputs = list(range(-n_outputs, 0))
+
+        for i in range(1, num_stages):
+            _inputs = inputs[:i] + [inputs[i]] * 2
+
+            _kwargs = kwargs
+            _kwargs['upsample'] = _kwargs['dense']['upsample']
+            _kwargs['upsample']['filters'] = decoder_filters[-i:]
+            _kwargs['blocks']['filters'] = decoder_filters[-i:]
+
+            outputs = super().decoder(_inputs, name=name,
+                                      return_all=True, **{**_kwargs, 'num_stages': i})
+
+            for j, x in enumerate(inputs[:i]):
+                x = combine([x, outputs[::-1][j]], **combine_args)
+
+                dense_args = {**unpack_args(kwargs['dense'], j, num_stages), **kwargs}
+                x = conv_block(x, name='x-{}-{}'.format(j, i), **dense_args)
+
+                if j == 0:
+                    semantic_outputs.append(x)
+
+                inputs[j] = combine([x, inputs[j]], name='concat-{}-{}'.format(j, i-j), **combine_args)
+
+        kwargs['upsample']['filters'] = decoder_filters
+        kwargs['blocks']['filters'] = decoder_filters
+
+        x = super().decoder(inputs, name=name, filters=decoder_filters,
+                            return_all=False, **{**kwargs, 'num_stages': num_stages})
+
+        semantic_outputs.append(x)
+
+        return [semantic_outputs[i] for i in n_outputs]
+
+    @classmethod
+    def head(cls, inputs, targets, name='head', **kwargs):
+        res = []
+        for i, x in enumerate(inputs):
+            res.append(super().head(x, targets, name=name+'-'+str(i), **kwargs))
+        return res
```

Comment:
What will happen to this list afterwards?

Reply:
It will be used to create outputs